witness/golang/internal/persistence/inmemory/inmemory.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2022 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package inmemory provides a persistence implementation that lives only in memory.
+package inmemory
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/google/trillian-examples/witness/golang/internal/persistence"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// NewInMemoryPersistence returns a persistence object that lives only in memory.
+func NewInMemoryPersistence() persistence.LogStatePersistence {
+	return &inMemoryPersistence{
+		checkpoints: make(map[string]checkpointState),
+	}
+}
+
+type checkpointState struct {
+	rawChkpt     []byte
+	compactRange []byte
+}
+
+type inMemoryPersistence struct {
+	mu          sync.RWMutex
+	checkpoints map[string]checkpointState
+}
+
+func (p *inMemoryPersistence) Init() error {
+	return nil
+}
+
+func (p *inMemoryPersistence) Logs() ([]string, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	res := make([]string, 0, len(p.checkpoints))
+	for k := range p.checkpoints {
+		res = append(res, k)
+	}
+	return res, nil
+}
+
+func (p *inMemoryPersistence) ReadOps(logID string) (persistence.LogStateReadOps, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	var cp *checkpointState
+	if got, ok := p.checkpoints[logID]; ok {
+		cp = &got
+	}
+	return &readWriter{
+		read: cp,
+	}, nil
+}
+
+func (p *inMemoryPersistence) WriteOps(logID string) (persistence.LogStateWriteOps, error) {
+	p.mu.RLock()
```

Comment:
This could then be `Lock` and `Unlock` here, and I think you wouldn't need the `expectAndWrite` anymore

Reply:
As replied above, this wouldn't work for the Reader as it doesn't have Close.

That aside, I still don't like this approach as it's effectively a mutex around the whole of the witness operation. The current implementation allows multiple log checkpoint updates to be processed concurrently. If these are all different logs then this is really quite performant.

Maybe I'm not seeing the wood for the trees because I wrote this code, but it's interesting that both you and Sarah seem to respond to this locking like it's totally whacky. To me, this looks pretty standard. What am I missing?